dot/parachain/backing/active_leaves_update.go

```diff
@@ -0,0 +1,478 @@
+				assignment = &coreParaID
+			}
+			groups[coreParaID] = validatorIndexes
+		}
+	}
+
+	tableContext := TableContext{
+		validator:  localValidator,
+		groups:     groups,
+		validators: validators,
+	}
+
+	tableConfig := Config{
+		AllowMultipleSeconded: mode.IsEnabled,
+	}
+
+	newPerRelayParentState := perRelayParentState{
+		prospectiveParachainsMode: mode,
+		relayParent:               relayParent,
+		assignment:                assignment,
+		table:                     newTable(tableConfig),
+		tableContext:              tableContext,
+		fallbacks:                 make(map[parachaintypes.CandidateHash]attestingData),
+		awaitingValidation:        make(map[parachaintypes.CandidateHash]bool),
+		issuedStatements:          make(map[parachaintypes.CandidateHash]bool),
+		backed:                    make(map[parachaintypes.CandidateHash]bool),
+		minBackingVotes:           minBackingVotes,
+	}
+
+	return &newPerRelayParentState, nil
+}
+
+func minBackingVotes(rt runtime.Instance) (uint32, error) {
+	votes, err := rt.ParachainHostMinimumBackingVotes()
+	if err != nil && errors.Is(err, wazero_runtime.ErrExportFunctionNotFound) {
+		logger.Tracef(
+			"%s is not supported by the current Runtime API",
+			runtime.ParachainHostMinimumBackingVotes,
+		)
+		return LEGACY_MIN_BACKING_VOTES, nil
+	}
+	return votes, err
+}
+
+func fetchParachainHostData(rt runtime.Instance) (
+	parachaintypes.SessionIndex,
+	[]parachaintypes.ValidatorID,
+	parachaintypes.ValidatorGroups,
+	scale.VaryingDataTypeSlice,
+	bool, // bool is used to indicate if all data was fetched successfully.
+) {
+	var (
+		sessionIndex    parachaintypes.SessionIndex
+		validators      []parachaintypes.ValidatorID
+		validatorGroups *parachaintypes.ValidatorGroups
+		cores           *scale.VaryingDataTypeSlice
+	)
+
+	// Error channel to receive errors from goroutines.
+	errCh := make(chan error, 4)
+
+	// WaitGroup to wait for all goroutines to finish.
+	var wg sync.WaitGroup
+
+	wg.Add(4)
+
+	go func() {
+		defer wg.Done()
+
+		var err error
+		sessionIndex, err = rt.ParachainHostSessionIndexForChild()
+		if err != nil {
+			errCh <- fmt.Errorf("getting session index: %w", err)
+			return
+		}
+	}()
+
+	go func() {
+		defer wg.Done()
+
+		var err error
+		validators, err = rt.ParachainHostValidators()
+		if err != nil {
+			errCh <- fmt.Errorf("getting validators: %w", err)
+			return
+		}
+	}()
+
+	go func() {
+		defer wg.Done()
+
+		var err error
+		validatorGroups, err = rt.ParachainHostValidatorGroups()
+		if err != nil {
+			errCh <- fmt.Errorf("getting validator groups: %w", err)
+			return
+		}
+	}()
+
+	go func() {
+		defer wg.Done()
+
+		var err error
+		cores, err = rt.ParachainHostAvailabilityCores()
+		if err != nil {
+			errCh <- fmt.Errorf("getting availability cores: %w", err)
+			return
+		}
+	}()
+
+	wg.Wait()
+
+	if len(errCh) > 0 {
+		for err := range errCh {
+			logger.Error(err.Error())
```

Comment:
You could use `errors.Join` to join these errors and  return the resulting error.